ansible_base/jwt_consumer/common/auth.py

```diff
@@ -0,0 +1,203 @@
+            self.log_and_raise(f"The specified file {file_path} is not a file")
+        except PermissionError:
+            self.log_and_raise(f"Permission error when reading {file_path}")
+        except Exception as e:
+            self.log_and_raise(f"Failed reading {file_path}: {e}")
+
+    def get_decryption_key(self, url_or_string, **kwargs):
+        timeout = kwargs.get('timeout', 30)
+        validate_certs = kwargs.get('validate_certs', True)
+        url_info = urlparse(url_or_string)
+        key = None
+        logger.info(f"Loading decryption key from {url_or_string} scheme {url_info.scheme}")
+        if url_info.scheme in ["http", "https"]:
+            key = self.get_decryption_key_from_url(url_or_string, timeout, validate_certs)
+        elif url_info.scheme == "file":
+            file_path = url_info.path
+            key = self.get_decryption_key_from_file(file_path)
+        elif url_info.scheme == "" and url_info.path != "":
+            logger.debug("Assuming decryption key is the actual cert")
+            key = url_or_string
+
+        if key is None:
+            self.log_and_raise(f"Unable to determine how to handle {url_or_string} to get key")
+        elif not key.startswith('-----BEGIN PUBLIC KEY-----') and not key.endswith('-----END PUBLIC KEY-----'):
+            logger.debug(key)
+            self.log_and_raise("Returned key does not start and end with BEGIN/END PUBLIC KEY")
+        logger.info("Decryption key appears valid")
+        logger.debug(f"{key}")
+        return key
+
+    def map_user_fields(self, user, token):
+        user_needs_save = False
+        for attribute in self.mapped_user_fields:
+            old_value = getattr(user, attribute, None)
+            new_value = token.get(attribute, None)
+            if old_value != new_value:
+                logger.debug(f"Changing {attribute} for {user.username} from {old_value} to {new_value}")
+                setattr(user, attribute, new_value)
+                user_needs_save = True
+        if user_needs_save:
+            logger.info(f"Saving user {user.username}")
+            user.save()
+
+    def validate_token(self, token, decryption_key):
+        validated_body = None
+        required_fields = self.mapped_user_fields
+
+        # Ensure all of the internal pieces are part of the token but we don't want them in required_feidls or we will attempt to make them into the user object
+        local_required_field = ["sub", "claims", "exp"]
+        for field in required_fields:
+            if field not in local_required_field:
+                local_required_field.append(field)
+
+        # Decrypt the token
+        try:
+            logger.info("Decrypting token")
+            validated_body = jwt.decode(
+                token,
+                decryption_key,
+                audience="ansible-services",
+                options={"require": local_required_field},
+                issuer="ansible-issuer",
```

Comment:
I had to change hard-coded gateway values to match in order to pass the jwt verification phase. Maybe these should be configurable?